### ISSUE TITLE
What theme color should be used when a PWA window is out of scope?

### ISSUE THREAD

**Issue by alancutter** [w3c/manifest#755]
https://www.w3.org/TR/appmanifest/#theme_color-member says

> However, a document may override the default theme color through the inclusion of a valid [HTML] meta element whose name attribute is "theme-color".

I see this as a way for PWAs to dynamically enable dark mode for their site and have the window frame update as well. I don't think it makes sense for an out of scope page to override the app's choice of theme colour though.

What should the theme colour be when out of scope?
A: Manifest theme_color
B: Last used in-scope theme colour
C: Out of scope page theme colour

This question was first raised in: https://bugs.chromium.org/p/chromium/issues/detail?id=872121

**Comment by kenchris** [comment]
Out of scope pages might have their own theme-color meta tag as well

**Comment by dominickng** [comment]
@kenchris yes, but on desktop, we don't really think it's appropriate for an out-of-scope page's custom theme color to override the theme color of the original app window if the out-of-scope page is being opened in an app window.

I think I like the idea of (A) here - falling back to the manifest theme color when out of scope OR when in-scope and the meta theme-color isn't specified.

**Comment by kenchris** [comment]
Don't you think that some PWAs would like to indicate that what is being loaded it out of scope? and maybe have a `external_theme_color` or similar?

**Comment by dominickng** [comment]
Possibly, but the spec already [strongly recommends](https://www.w3.org/TR/appmanifest/#navigation-scope) for browsers do that on behalf of the site.

**Comment by g-ortuno** [comment]
What about B? When in dark mode, Twitter uses the page theme color to set the frame color. If the user goes out of scope I think it should still use the dark theme color.

**Comment by alancutter** [comment]
I think A is the most semantically sensible thing to do. I think it's reasonable for sites to deliver an updated manifest based on user preferences and push the change to the user's app. The dynamic theme_color for dark mode can help inform how we handle [manifest updates](https://www.w3.org/TR/appmanifest/#updating).

**Comment by marcoscaceres** [comment]
We need to clarify that it's left to the browser.

**Comment by kenchris** [comment]
There is a default theme color in browsers for when none is defined in the manifest or in meta tag. I would say that if the out of scope url has a meta tag for theme-color, use that, and if not, use the default browser theme color

**Comment by christianliebel** [comment]
> There is a default theme color in browsers for when none is defined in the manifest or in meta tag. I would say that if the out of scope url has a meta tag for theme-color, use that, and if not, use the default browser theme color

@kenchris In #879, I used your proposal as a basis for discussion. Looking forward to your feedback! .cc @marcoscaceres

**Comment by mgiuca** [comment]
As I wrote on #879, I don't think C is reasonable (I agree with @dominickng and @alancutter here), which is what was proposed on the PR. I don't think B is good either, it means your window is stateful, based on your navigation history. Strong preference for A.

I don't feel strongly that this requires a spec change. If anything, it just requires a clarification of what a manifest being "[applied](https://www.w3.org/TR/appmanifest/#dfn-apply)" actually means.

The spec seems to have conflicting ideas about whether the manifest is still [applied](https://www.w3.org/TR/appmanifest/#dfn-apply) to off-scope navigations. The definition of "[applied](https://www.w3.org/TR/appmanifest/#dfn-apply)" itself says what it means to be applied, but not when:

> A [manifest](https://www.w3.org/TR/appmanifest/#dfn-manifest) is **_applied_** to a [top-level browsing context](https://html.spec.whatwg.org/multipage/browsers.html#top-level-browsing-context), meaning that the members of the [manifest](https://www.w3.org/TR/appmanifest/#dfn-manifest) are affecting the presentation or behavior of a browsing context.

In fact, there is nowhere that outright states under what conditions a manifest is applied to a top-level browsing context, but there is this problematic definition of [navigation scope](https://www.w3.org/TR/appmanifest/#dfn-navigation-scope):

> A _**navigation scope**_ is a [URL](https://url.spec.whatwg.org/#concept-url) that represents the set of URLs to which an [application context](https://www.w3.org/TR/appmanifest/#dfn-application-context) can be navigated while the manifest is [applied](https://www.w3.org/TR/appmanifest/#dfn-apply).

That implies that it's impossible to navigate to an off-scope URL whilst the manifest is still being applied. However, the opposite is implied by this non-normative note:

> Nothing prevents an [application context](https://www.w3.org/TR/appmanifest/#dfn-application-context) from navigating to a [URL](https://url.spec.whatwg.org/#concept-url) that is outside of the application's [navigation scope](https://www.w3.org/TR/appmanifest/#dfn-navigation-scope), while still having the [manifest](https://www.w3.org/TR/appmanifest/#dfn-manifest) [applied](https://www.w3.org/TR/appmanifest/#dfn-apply) to it.

By that description, the manifest is still "[applied](https://www.w3.org/TR/appmanifest/#dfn-apply)" even when displaying an off-scope URL. It then follows that:

> the [default theme color](https://www.w3.org/TR/appmanifest/#dfn-default-theme-color) ... serves as the [theme color](https://www.w3.org/TR/appmanifest/#dfn-theme-color) for all browsing contexts to which the manifest is [applied](https://www.w3.org/TR/appmanifest/#dfn-apply).

If we interpret it as that the manifest is still [applied](https://www.w3.org/TR/appmanifest/#dfn-apply) on an off-scope URL, then A is already the case. If any work is to be done here, I think it is clarifying normatively that a manifest is still [applied](https://www.w3.org/TR/appmanifest/#dfn-apply) even to an off-scope URL, which would involve changing the definition of [navigation scope]((https://www.w3.org/TR/appmanifest/#dfn-navigation-scope)).

**Comment by mgiuca** [comment]
I uploaded #880 which addresses the three problems I identified above:

* Navigation scope should not be defined as "the URLs the manifest can be applied to" (since it can apply to any URL off-scope).
* Explicitly state when a user agent should "apply" a manifest.
* A document's theme color MAY override the default theme color, if it is in scope, otherwise it SHOULD NOT override the default theme color.
